### ISSUE TITLE
[Discussion] Enhancing Youtube Heuristics 

### ISSUE THREAD

**Issue by Demizo** [nukeop/nuclear#1381]
TLDR; I am proposing weight changes to the Youtube heuristics in order to massively improve Nuclear's ability to fetch the right track. I wanted to discuss these changes before implementing anything final.

Currently Nuclear often grabs the wrong track off Youtube when playing a song. This happens regardless of genre or song popularity. Earlier this year I contributed to a similar streaming program called SpotTube which had the same issue. Like Nuclear, SpotTube had built in heuristics to validate the accuracy of songs and weed out things like live videos; however, its track matching was still very unreliable. After working on the heuristics, I found that the most reliable way to ensure the correct song was to place emphasis on matching song duration. This change greatly improved SpotTube's ability to fetch the correct tracks.

I recently switched to Nuclear and want to help improve this app's heuristics as well. I am aware that a heuristics overhaul was discussed in the past and is possibly in the works on a separate branch, but I want to propose easy fixes that can at least improve the track matching temporarily.

Here's the changes I made to the heuristics to improve track matching:

[DISCLAIMER] These changes have not been thoroughly tested, ideally I would like to create a large and varied test bench (with varying popularity and genre) to validate this and future heuristics against... 
(Forgive the styling changes to the code, that was accidental)

1. Adding "VEVO" and "Video" to peneilized words list. This limits the amount of music videos when possible...
```
const penalizedWords = ["remix", "full album", "album", "VEVO", "video"];
```

2. Changed "trackTitle" to "title" for the Title Score and Verbatim Substring Score. "trackTitle" includes the Artist name and the dash "-" this meant that incorrect songs from artists that followed this format were sometimes picked over correct tracks that only contained the song name. The verbatim substring score was also doubled in weight to once again place emphasis on the right song name from an artist.
```
    const trackTitle = YoutubeHeuristics.createTitle({ artist, title });
    const lowercaseResultTitle = track.title.toLowerCase();

    const titleScore = (1 - levenshtein.get(title, lowercaseResultTitle) / title.length) * 100;

    const verbatimSubstringScore = lowercaseResultTitle.includes(title) ? 200 : 0;
```

3. The influence of duration accuracy was increased by 8x. Duration is a reliable fingerprint for a song and can root out things like remixes, music videos, and different songs.
```
    const durationScore = track.duration && duration ? (1 - durationDelta / duration) * 800 : 0;
```

4. Penalized words were given a slightly higher weight since they are more important than promoted words. An incorrect song having "Official" in its name often meant it got picked. Penalized words should carry more weight than promoted ones.
```
const penalizedWordsScore = penalizedWords.some((word) => lowercaseResultTitle.includes(word))  ? 0 : 200;
```

5. To balance out the other changes, the weight of channel name was increased to restore its ability to provide accurate artists.
```
    const channelNameScore = track.author.name.toLowerCase().includes(artist.toLowerCase())  ? 200  : 0;
```

With these changes I was able to massively improve the accuracy of the heuristics; however, much more testing will be needed to ensure it can be the best it can be. Thank you for taking the time to read this, Nuclear is a fantastic tool that I would love to see grow!

**Comment by nukeop** [comment]
Thanks for your tips, there's no real science to the heuristics I came up with, and it was all what made sense at the time.

Ad. 1 - Penalizing music videos should not be necessary - we've got sponsorblock that will cut out non-music segments so that you're left only with the music.

Ad. 2 - This could help fix some problems where a correct artist is selected, but it still chooses the wrong track.

Ad. 3 - Sounds good, only needs an illustrative test.

Ad. 4  - Sounds sensible and I would like to add this

Ad. 5 - Necessary if we include 3 and 4.

**Comment by Demizo** [comment]
Good point on the penalized words. I will say that while sponsor block does often work just fine, some music videos are a variation of the original song. However, the other proposed changes would probably root out music videos that do vary from the original song. 

I have always wanted to look more into code testing so I will see if I can find a good way to validate these changes as well as potential performance impacts.

Thanks again for the input! Great app!

**Comment by mehrsharma** [comment]
hi! is anyone working on this? I'm new to the project and I could pick it up

**Comment by Demizo** [comment]
@mehrsharma I've made some progress on improving this issue, and have it to the point where it is seemingly always right. However, it still needs actual testing and validation. I have very little experience writing tests so I wouldn't mind a helping hand. I have a fork of nuclear with a branch containing my heuristics fixes (the changes haven't been committed yet I can do that later today), if you want to take a look at what I've done.

**Comment by nukeop** [comment]
Sure, feel free to open a pull request and I can help with the tests (check the option to let me commit to your branch).

**Comment by mehrsharma** [comment]
@Demizo I looked over the existing tests, although I'm not too sure about what modifications/new tests would need to be added for the new heuristics changes. I could definitely help out given some more guidance though!

**Comment by Demizo** [comment]
Thanks, I added my current changes to a draft pull request. Currently it also contains a bandaid fix for Nuclear fetching "undefined" tracks. This is a bug in the current release of Nuclear and should have a separate bug report / pull request to fix it.

**Comment by nukeop** [comment]
Implemented
